The entry point Main method should not be allowed to return int by-ref because it is not supported by CLR

**Version Used**: 

Microsoft Visual Studio Enterprise 2017
Version 15.0.26228.9 D15RTWSVC
Microsoft .NET Framework
Version 4.6.01586

**Steps to Reproduce**:

Compile as a console application and run:
```
using System;

class C
{
    static ref int Main()
    {
        throw new Exception();
    }
}
```

**Expected Behavior**: A compile-time error: error CS5001: Program does not contain a static 'Main' method suitable for an entry point.

**Actual Behavior**: No compile-time errors. Unhandled exception at runtime: System.MethodAccessException: Entry point must have a return type of void, integer, or unsigned integer.

--- a/src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs
@@ -613,6 +613,11 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
                 return false;
             }
 
+            if (RefKind != RefKind.None)
+            {
+                return false;
+            }
+
             if (Parameters.Length == 0)
             {
                 return true;

--- a/src/Compilers/CSharp/Test/Emit/Emit/EntryPointTests.cs
+++ b/src/Compilers/CSharp/Test/Emit/Emit/EntryPointTests.cs
@@ -1444,5 +1444,45 @@ class Program
                 Diagnostic(ErrorCode.ERR_MultipleEntryPoints, "Main").WithLocation(4, 24)
                 );
         }
+
+        [WorkItem(17923, "https://github.com/dotnet/roslyn/issues/17923")]
+        [Fact]
+        [CompilerTrait(CompilerFeature.RefLocalsReturns)]
+        public void RefIntReturnMainEmpty()
+        {
+            var source = @"
+class Program
+{
+    public static ref int Main() { throw new System.Exception(); }
+}";
+
+            var compilation = CreateCompilationWithMscorlib(source, options: TestOptions.DebugExe);
+            compilation.VerifyDiagnostics(
+                // (4,27): warning CS0028: 'Program.Main()' has the wrong signature to be an entry point
+                //     public static ref int Main() {}
+                Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("Program.Main()").WithLocation(4, 27),
+                // error CS5001: Program does not contain a static 'Main' method suitable for an entry point
+                Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1));
+        }
+
+        [WorkItem(17923, "https://github.com/dotnet/roslyn/issues/17923")]
+        [Fact]
+        [CompilerTrait(CompilerFeature.RefLocalsReturns)]
+        public void RefIntReturnMainWithParams()
+        {
+            var source = @"
+class Program
+{
+    public static ref int Main(string[] args) { throw new System.Exception(); }
+}";
+
+            var compilation = CreateCompilationWithMscorlib(source, options: TestOptions.DebugExe);
+            compilation.VerifyDiagnostics(
+                // (4,27): warning CS0028: 'Program.Main(string[])' has the wrong signature to be an entry point
+                //     public static ref int Main(string[] args) { throw new System.Exception(); }
+                Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("Program.Main(string[])"),
+                // error CS5001: Program does not contain a static 'Main' method suitable for an entry point
+                Diagnostic(ErrorCode.ERR_NoEntryPoint));
+        }
     }
 }

